src/core/unif.ml

```diff
@@ -36,81 +36,59 @@ let no_problems : problems =
-  let replace_nl_var x =
-    if VarSet.mem x !nl_vars then Bindlib.new_var mkfree "_" else x
-  in
-  try Some (Array.map replace_nl_var (Array.map to_var ts))
-  with Not_a_var -> None
-
-(** [instantiation m ts u] checks whether, in a problem [m[ts]=u], [m] can be
-   instantiated and returns the corresponding instantiation. It does not check
-   whether the instantiation is closed though. *)
-let instantiation : meta -> term array -> term -> tmbinder Bindlib.box option
-  = fun m ts u ->
+(** [instantiation ctx m ts u] checks whether, in a problem [m[ts]=u], [m] can
+    be instantiated and returns the corresponding instantiation. It does not
+    check whether the instantiation is closed though. *)
+let instantiation : ctxt -> meta -> term array -> term ->
+  tmbinder Bindlib.box option = fun ctx m ts u ->
   if (!can_instantiate || internal m) && not (occurs m u) then
-    match nl_distinct_vars ts with
-    | None -> None
-    | Some vs -> Some (Bindlib.bind_mvar vs (lift u))
+    match nl_distinct_vars ctx ts with
+    | None     -> None
+    | Some(vs) -> Some (Bindlib.bind_mvar vs (lift u))
   else None
 
-(** [instantiate m ts u] check whether, in a problem [m[ts]=u], [m] can be
-   instantiated and, if so, instantiate it. *)
-let instantiate : meta -> term array -> term -> bool = fun m ts u ->
-  match instantiation m ts u with
-  | Some bu when Bindlib.is_closed bu -> set_meta m (Bindlib.unbox bu); true
+(** [instantiate ctx m ts u] check whether, in a problem [m[ts]=u], [m] can be
+    instantiated and, if so, instantiate it. *)
+let instantiate : ctxt -> meta -> term array -> term -> bool =
+  fun ctx m ts u ->
+  match instantiation ctx m ts u with
+  | Some(bu) when Bindlib.is_closed bu -> set_meta m (Bindlib.unbox bu); true
   | _ -> false
 
 (** [solve cfg p] tries to solve the unification problems of [p] and
-   returns the constraints that could not be solved. *)
+    returns the constraints that could not be solved. *)
 let rec solve : problems -> unif_constrs = fun p ->
-  (*if !log_enabled then log_unif "%a" pp_problem p;*)
   match p with
   | { to_solve = []; unsolved = []; _ } -> []
   | { to_solve = []; unsolved = cs; recompute = true } ->
      solve {no_problems with to_solve = cs}
   | { to_solve = []; unsolved = cs; _ } -> cs
-  | { to_solve = (t,u)::to_solve; _ } -> solve_aux t u {p with to_solve}
+  | { to_solve = (c,t,u)::to_solve; _ } -> solve_aux c t u {p with to_solve}
 
 (** [solve_aux t1 t2 p] tries to solve the unificaton problem given by [p] and
     the constraint [(t1,t2)], starting with the latter. *)
-and solve_aux : term -> term -> problems -> unif_constrs = fun t1 t2 p ->
-  let (h1, ts1) = Eval.whnf_stk t1 []
-  and (h2, ts2) = Eval.whnf_stk t2 [] in
-  if !log_enabled then
-    log_unif "solve %a ≡ %a" pp (add_args h1 ts1) pp (add_args h2 ts2);
+and solve_aux : ctxt -> term -> term -> problems -> unif_constrs =
+  fun ctx t1 t2 p ->
+  let t1 = Eval.whnf ctx t1 in
+  let t2 = Eval.whnf ctx t2 in
```

Comment:
From this point, the context ctx is completely useless because defined variables of ctx have already been unfolded by Eval.whnf in lines 72-73. This means that the ctxt argument of solve_aux could be completely removed by computing the whnf in solve by having in line 66 | { to_solve = (c,t,u)::to_solve; _ } -> solve_aux (Eval.whnf c t) (Eval.whnf c u) {p with to_solve}.

Reply:
So, after all, I wonder if it is really a good idea that whnf_stk does not take a ctxt argument. @rlepigre , @gabrielhdt , what do you think?

Reply:
Regarding the context of `solve_aux`, there might be variables not unfolded by `whnf` needed, unless we only call back `solve` recursively on `t1` and `t2` (which are in whnf).

Regarding the `whnf_stk`, it seems to me that not taking a context would slow down `whnf` by "uselessly" trying to build `llet`, but anyway we do not use contexts often (for now). On the other hand, adding an argument to `whnf_stk` might slow the reduction engine as well (though I'm not sure), so I don't know.

Reply:
I'm a bit worried about performances of the reduction engine, but maybe that would not be too bad. The only way to know for sure is to try, but I don't know if this is a good idea to attempt this in this here. Maybe in a follow-up PR?

Reply:
Gabriel said: "Regarding the context of solve_aux, there might be variables not unfolded by whnf needed, unless we only call back solve recursively on t1 and t2 (or h1 and h2) (which are in whnf)." I don't understand. In the current definition of whnf, the first thing that is done to build lets from the context and then reduce them. Hence, after whnf, all the defined variables of the context have been unfolded. Don't you agree?
Concerning efficiency, that's true that having one argument less may be better. On the other hand, if we have a context argument, then we can unfold context variables lazily by modifying the definition of eq_modulo. Moreover, it seems to me that it would then be easier to introduce some sharing in the unfolding of context variables.

Reply:
Ah indeed I forgot to consider the `let`s on top. And I definitely agree with you regarding laziness and sharing and ease of development. So adding back the ctxt in `whnf_stk` seems to be the way to go.

Reply:
It also implies giving the context to `tree_walk` since it might call `snf` and is highly likely to call `whnf`

Reply:
Do you want to give it a try?

Reply:
If everything is fine, we could merge now.

Reply:
I just pushed the new `whnf` in 75aacb6

Reply:
Cool, it is true that it may very well not be too bad for efficiency. Carrying one extra argument is not such a big deal in OCaml. At least if there are enough registers for all the arguments.